### PR TITLE
[BUGFIX] Allow for RuntimeDataConnector to accept custom query while suppressing temp table creation

### PR DESCRIPTION
**Pull request by NathanFarmer**
Changes proposed in this pull request:
- Handle methods for getting column names as well as the table head when a custom query is used
- https://github.com/great-expectations/great_expectations/issues/3153

Previous Design Review notes:
- In Design Review it was suggested to transform the provided query object (`<class 'sqlalchemy.sql.elements.TextClause'>`) by wrapping with `sa.select()`. This would be handled in `great_expectations/execution_engine/sqlalchemy_batch_data.py:165`.
- The suggestion above results in a `<class 'sqlalchemy.sql.selectable.Select'>` object which has no attributes `name` or `schema`, failing to resolve the error encountered by the user. Futhermore, the custom query object is used elsewhere, such as in `great_expectations/expectations/metrics/table_metrics/table_head.py` and using the `sa.select()` wrapper does not resolve additional errors encountered there.
- In the typical pattern (without providing a custom query) `selectable` is provided as `<class 'sqlalchemy.sql.schema.Table'>`. This object fails to capture the various ways a custom query could be used in this context.
- The initial bug fix of handling the `<class 'sqlalchemy.sql.elements.TextClause'>` downstream of `sqlalchemy_batch_data.py` is what is proposed in this PR. Semantically, [the SQL Alchemy creators state](https://docs.sqlalchemy.org/en/14/core/selectable.html) that "[t]he term “selectable” refers to any object that rows can be selected from." The `TextClause` object meets this definition even if it is not descended from `FromClause` as the docs go on to describe.

### Definition of Done

- [X] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [X] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: great_expectations/expectations/metrics/table_metrics/table_column_types.py [hunk 1]
```diff
@@ -83,8 +88,13 @@ def _spark(
...
 def _get_sqlalchemy_column_metadata(engine, batch_data: SqlAlchemyBatchData):
-    table_selectable = batch_data.source_table_name or batch_data.selectable.name
-    schema_name = batch_data.source_schema_name or batch_data.selectable.schema
+    # if a custom query was passed
+    if isinstance(batch_data.selectable, TextClause):
+        table_selectable = batch_data.selectable
+        schema_name = None
+    else:
+        table_selectable = batch_data.source_table_name or batch_data.selectable.name
+        schema_name = batch_data.source_schema_name or batch_data.selectable.schema
```
Comment (talagluck) [inline great_expectations/expectations/metrics/table_metrics/table_column_types.py]: Could we add type hints here to make this clearer? Is the `table_selectable` on L96 generally a consistent type?

**Review by talagluck — COMMENTED** [review]
Overall this looks great! One suggestion for type hints to make this a bit easier to read. Would it also be possible to add a couple of unit tests to demonstrate that this works as expected?

File: great_expectations/expectations/metrics/table_metrics/table_column_types.py (hunk 1 above)
Comment (NathanFarmer) [reply great_expectations/expectations/metrics/table_metrics/table_column_types.py]: Yes, that makes sense. Typically `table_selectable` would be a string representing the name of the table. I will make this change.

**Review by talagluck — APPROVED** [review]
LGTM! This is awesome!

**Comment by cohenbenjamin05** [comment]
I am experiencing an issue with this after the bugfix. I think somewhere in the code it is still expecting a temp table to be used as the syntax of the query GE is trying to execute is incorrect. The query it is trying to run would be fine if it was querying a temp table but instead it's trying to execute the following query: `SELECT count(*) AS "table.row_count" 
FROM select * from schema.table`. See below for the stack trace (where my custom sql is just `select * from schema.table`).

```
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1772, in _execute_context
    cursor, statement, parameters, context
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 717, in do_execute
    cursor.execute(statement, parameters)
psycopg2.errors.SyntaxError: syntax error at or near "select"
LINE 2: FROM select * from schema.table
             ^


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/great_expectations/validations/reporting.py", line 114, in <module>
    main()
  File "/great_expectations/validations/reporting.py", line 101, in main
    LOG.info(my_validator.validate())
  File "/usr/local/lib/python3.7/site-packages/great_expectations/validator/validator.py", line 1218, in validate
    "result_format": result_format,
  File "/usr/local/lib/python3.7/site-packages/great_expectations/validator/validator.py", line 472, in graph_validate
    metrics = self.resolve_validation_graph(graph, metrics, runtime_configuration)
  File "/usr/local/lib/python3.7/site-packages/great_expectations/validator/validator.py", line 517, in resolve_validation_graph
    runtime_configuration=runtime_configuration,
  File "/usr/local/lib/python3.7/site-packages/great_expectations/validator/validator.py", line 558, in _resolve_metrics
    metrics_to_resolve, metrics, runtime_configuration
  File "/usr/local/lib/python3.7/site-packages/great_expectations/execution_engine/execution_engine.py", line 295, in resolve_metrics
    resolved_metrics.update(self.resolve_metric_bundle(metric_fn_bundle))
  File "/usr/local/lib/python3.7/site-packages/great_expectations/execution_engine/sqlalchemy_execution_engine.py", line 748, in resolve_metric_bundle
    sa.select(query["select"]).select_from(selectable)
  File "<string>", line 2, in execute
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/util/deprecations.py", line 390, in warned
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 3108, in execute
    return connection.execute(statement, *multiparams, **params)
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1263, in execute
    return meth(self, multiparams, params, _EMPTY_EXECUTION_OPTS)
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/sql/elements.py", line 324, in _execute_on_connection
    self, multiparams, params, execution_options
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1462, in _execute_clauseelement
    cache_hit=cache_hit,
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1815, in _execute_context
    e, statement, parameters, cursor, context
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1996, in _handle_dbapi_exception
    sqlalchemy_exception, with_traceback=exc_info[2], from_=e
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1772, in _execute_context
    cursor, statement, parameters, context
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 717, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.ProgrammingError: (psycopg2.errors.SyntaxError) syntax error at or near "select"
LINE 2: FROM select * from schema.table
             ^

[SQL: SELECT count(*) AS "table.row_count" 
FROM select * from schema.table]
(Background on this error at: https://sqlalche.me/e/14/f405 )
```

**Comment by NathanFarmer** [comment]
Hi @cohenbenjamin05, thanks for pointing this out! Would you mind opening a new issue so that we can get this in our queue? Please also include which SQL Alchemy dialect you are using in this example.

**Comment by cohenbenjamin05** [comment]
Will do, thanks for the quick response!

**Comment by cohenbenjamin05** [comment]
@NathanFarmer - Here is the issue I created https://github.com/great-expectations/great_expectations/issues/3398. Let me know if you need any additional information. Thanks for the help!

**Pull request merged**